qiskit/algorithms/evolvers/pvqd/pvqd.py

```diff
@@ -0,0 +1,390 @@
+        trotterized.append(
+            PauliEvolutionGate(hamiltonian, time=dt, synthesis=self.evolution), ansatz.qubits
+        )
+
+        # define the overlap of the Trotterized state and the ansatz
+        x = ParameterVector("w", self.ansatz.num_parameters)
+        shifted = ansatz.assign_parameters(current_parameters + x)
+        overlap = StateFn(trotterized).adjoint() @ StateFn(shifted)
+
+        converted = self.expectation.convert(overlap)
+
+        def evaluate_loss(
+            displacement: Union[np.ndarray, List[np.ndarray]]
+        ) -> Union[float, List[float]]:
+            """Evaluate the overlap of the ansatz with the Trotterized evolution.
+
+            Args:
+                displacement: The parameters for the ansatz.
+
+            Returns:
+                The fidelity of the ansatz with parameters ``theta`` and the Trotterized evolution.
+            """
+            if isinstance(displacement, list):
+                displacement = np.asarray(displacement)
+                value_dict = {x_i: displacement[:, i].tolist() for i, x_i in enumerate(x)}
+            else:
+                value_dict = dict(zip(x, displacement))
+
+            sampled = self._sampler.convert(converted, params=value_dict)
+
+            # in principle we could add different loss functions here, but we're currently
+            # not aware of a use-case for a different one than in the paper
+            return 1 - np.abs(sampled.eval()) ** 2
+
+        if _is_gradient_supported(ansatz) and self.gradients:
+
+            def evaluate_gradient(displacement: np.ndarray) -> np.ndarray:
+                """Evaluate the gradient with the parameter-shift rule.
+
+                This is hardcoded here since the gradient framework does not support computing
+                gradients for overlaps.
+
+                Args:
+                    displacement: The parameters for the ansatz.
+
+                Returns:
+                    The gradient.
+                """
+                # construct lists where each element is shifted by plus (or minus) pi/2
+                dim = displacement.size
+                plus_shifts = (displacement + np.pi / 2 * np.identity(dim)).tolist()
+                minus_shifts = (displacement - np.pi / 2 * np.identity(dim)).tolist()
+
+                evaluated = evaluate_loss(plus_shifts + minus_shifts)
+
+                gradient = (evaluated[:dim] - evaluated[dim:]) / 2
+
+                return gradient
+
+        else:
+            evaluate_gradient = None
+
+        return evaluate_loss, evaluate_gradient
+
+    def evolve(self, evolution_problem: EvolutionProblem) -> EvolutionResult:
+        """
+        Args:
+            evolution_problem: The evolution problem containing the hamiltonian, total evolution
+                time and observables to evaluate. Note that :class:`~.PVQD` currently does not support
+                hamiltonians of type :class:`~.MatrixOp`.
```

Comment:
Do we check this and raise an error to that effect? I do not see any such logic. But I do see in looking further that this seems more to be an aspect of EvolutionProblem and the way thats handled than PVQD itself argably - at least thats what my take away from the unit test is. So here I would see a ValueError as a result of supplying a MatrixOp - a condition not noted in the raises.

Reply:
IIRC that's because the `PauliEvolutionGate` doesn't handle `MatrixOp`s 🙂 to support them, p-VQD would have to use exact matrix exponentiation. I could actually just add that, it's not a lot more effort than raising the error 😉

Reply:
Support for `MatrixOp` is added 👍🏻